nengo_dl/tests/test_converter.py

```diff
@@ -0,0 +1,460 @@
+
+    with pytest.warns(UserWarning, match="falling back to a TensorNode"):
+        conv = converter.Converter(model, allow_fallback=True)
+    conv.verify(training=False)
+    conv.verify(training=True)
+
+
+@pytest.mark.training
+def test_fallback(Simulator):
+    inp = x = tf.keras.Input(shape=(2, 2))
+
+    class MyLayer(tf.keras.layers.Layer):
+        def build(self, input_shapes):
+            super().build(input_shapes)
+            self.kernel = self.add_weight(
+                shape=(), initializer=tf.initializers.RandomUniform()
+            )
+
+        def call(self, inputs):
+            assert inputs.shape[1:] == (2, 2)
+            return tf.reshape(
+                inputs * tf.cast(self.kernel, inputs.dtype), shape=(-1, 4)
+            )
+
+    layer = MyLayer()
+    x = layer(x)
+    x = tf.keras.layers.Reshape((2, 2))(x)
+    x = layer(x)
+    x = tf.keras.layers.Reshape((2, 2))(x)
+    x = layer(x)
+
+    model = tf.keras.Model(inp, x)
+    conv = converter.Converter(model, allow_fallback=True)
+
+    with Simulator(conv.net) as sim:
+        # check that weights are being shared correctly
+        assert len(sim.keras_model.trainable_weights) == 1
+        assert sim.keras_model.trainable_weights[0].shape == ()
+
+    conv.verify(training=False)
+    conv.verify(training=True)
+
+
+def test_add():
+    inp = [tf.keras.Input(shape=(2, 2)), tf.keras.Input(shape=(2, 2))]
+    out = tf.keras.layers.Add()(inp)
+
+    _test_convert(inp, out)
+
+
+def test_average():
+    inp = [tf.keras.Input(shape=(2, 2)), tf.keras.Input(shape=(2, 2))]
+    out = tf.keras.layers.Average()(inp)
+
+    _test_convert(inp, out)
+
+
+def test_concatenate():
+    inp = [
+        tf.keras.Input(shape=(2, 3)),
+        tf.keras.Input(shape=(2, 3)),
+        tf.keras.Input(shape=(4, 3)),
+    ]
+    x = tf.keras.layers.Concatenate(axis=1)(inp[:2])
+    x = tf.keras.layers.Concatenate(axis=2)([x, inp[2]])
+
+    _test_convert(inp, x)
+
+
+def test_zero_padding():
+    inp = x = tf.keras.Input(shape=(14, 2))
+    x = tf.keras.layers.ZeroPadding1D(padding=1)(x)
+    x = tf.keras.layers.Reshape((4, 4, 2))(x)
+    x = tf.keras.layers.ZeroPadding2D(padding=((2, 0), (0, 4)))(x)
+    x = tf.keras.layers.Reshape((2, 2, 2, 12))(x)
+    x = tf.keras.layers.ZeroPadding3D(padding=(1, 0, 3))(x)
+    if utils.tf_gpu_installed:
+        x = tf.keras.layers.Reshape((12, 8, 8))(x)
+        x = tf.keras.layers.ZeroPadding2D(padding=1, data_format="channels_first")(x)
+
+    _test_convert(inp, x)
+
+
+def test_batch_normalization(rng):
+    inp = x = tf.keras.Input(shape=(4, 4))
+    x = tf.keras.layers.BatchNormalization(axis=1)(x)
+    x = tf.keras.layers.BatchNormalization(axis=2)(x)
+    x = tf.keras.layers.BatchNormalization(center=False, scale=False)(x)
+
+    model = tf.keras.Model(inputs=inp, outputs=x)
+
+    inp_vals = [rng.uniform(size=(2, 4, 4))]
+
+    # test using tensornode fallback
+    # TODO: there is some bug with using batchnormalization layers inside
```

Comment:
Here also, should we pull this into a separate issue to keep in the issues list / backlog?

Reply:
There's one here https://github.com/nengo/nengo-dl/issues/109